src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs

```diff
@@ -93,12 +96,30 @@ public void Contextualize(ViewContext viewContext)
                 throw new ArgumentNullException(nameof(viewContext));
             }
 
-            var baseName = viewContext.View.Path.Replace('/', '.').Replace('\\', '.');
-            if (baseName.StartsWith(".", StringComparison.OrdinalIgnoreCase))
+            // Given a view path "/Views/Home/Index.cshtml" we want a baseName like "MyApplication.Views.Home.Index"
+
+            var path = viewContext.ExecutingFilePath;
+
+            if (string.IsNullOrEmpty(path))
+            {
+                path = viewContext.View.Path;
+            }
+
+            // Trim the file extension from the end of the path
+            if (!string.IsNullOrEmpty(path) && Path.HasExtension(path))
```

Comment:
Might want to consider using a `StringBuilder` to do these replaces \ substrings if it happens in ordinary cases.